File: services/lucy-realtime-node/src/services/agora.service.js
const { RtcTokenBuilder, RtcRole } = require('agora-access-token');

async function createRtcToken(payload) {
  const appId = process.env.AGORA_APP_ID;
  const appCertificate = process.env.AGORA_APP_CERTIFICATE;
  const channelName = String(payload.channelName || payload.roomId || '').trim();
  const uid = String(payload.uid || payload.userId || '').trim();
  const expiresIn = Number(payload.expiresIn || process.env.AGORA_TOKEN_EXPIRES_IN || 3600);

  if (!appId || !appCertificate) {
    const error = new Error('Agora app id/certificate is not configured');
    error.statusCode = 500;
    throw error;
  }
  if (!channelName) {
    const error = new Error('channelName or roomId is required');
    error.statusCode = 400;
    throw error;
  }
  if (!uid) {
    const error = new Error('uid or userId is required');
    error.statusCode = 400;
    throw error;
  }

  const now = Math.floor(Date.now() / 1000);
  const privilegeExpireTime = now + expiresIn;
  const token = RtcTokenBuilder.buildTokenWithAccount(
    appId,
    appCertificate,
    channelName,
    uid,
    RtcRole.PUBLISHER,
    privilegeExpireTime,
  );

  return {
    appId,
    channelName,
    uid,
    token,
    expiresIn,
    expireAt: new Date(privilegeExpireTime * 1000).toISOString(),
  };
}

async function startRecording(roomId, payload) {
  return {
    roomId,
    recordingId: null,
    status: 'starting',
    options: payload,
  };
}

async function stopRecording(roomId, payload) {
  return {
    roomId,
    recordingId: payload.recordingId,
    status: 'stopping',
  };
}

module.exports = {
  createRtcToken,
  startRecording,
  stopRecording,
};

File: lucy-web-stack/lucy-backend/src/main/java/com/lucy/backend/realtime/service/AgoraService.java
package com.lucy.backend.realtime.service;

import com.lucy.backend.realtime.dto.TokenRequestDto;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Service
public class AgoraService {

    @Value("${lucy.agora.appId}")
    private String appId;

    @Value("${lucy.agora.appCertificate}")
    private String appCertificate;

    @Value("${lucy.agora.tokenExpiresIn}")
    private Integer defaultExpiresIn;

    public Map<String, Object> createRtcToken(TokenRequestDto payload) {
        String channelName = (payload.getChannelName() != null) ? payload.getChannelName().trim() : 
                             (payload.getRoomId() != null ? payload.getRoomId().trim() : "");
        String uid = (payload.getUid() != null) ? payload.getUid().trim() : 
                     (payload.getUserId() != null ? payload.getUserId().trim() : "");
        Integer expiresIn = payload.getExpiresIn() != null ? payload.getExpiresIn() : defaultExpiresIn;

        if (appId == null || appCertificate == null) {
            throw new RuntimeException("Agora app id/certificate is not configured");
        }
        if (channelName.isEmpty()) {
            throw new IllegalArgumentException("channelName or roomId is required");
        }
        if (uid.isEmpty()) {
            throw new IllegalArgumentException("uid or userId is required");
        }

        long now = Instant.now().getEpochSecond();
        long privilegeExpireTime = now + expiresIn;
        
        io.agora.media.RtcTokenBuilder tokenBuilder = new io.agora.media.RtcTokenBuilder();
        String token;
        try {
            int uidInt = Integer.parseInt(uid);
            token = tokenBuilder.buildTokenWithUid(appId, appCertificate, channelName, uidInt,
                    io.agora.media.RtcTokenBuilder.Role.Role_Publisher, (int) privilegeExpireTime);
        } catch (NumberFormatException e) {
            token = tokenBuilder.buildTokenWithUserAccount(appId, appCertificate, channelName, uid,
                    io.agora.media.RtcTokenBuilder.Role.Role_Publisher, (int) privilegeExpireTime);
        }

        // RTM Token
        io.agora.rtm.RtmTokenBuilder rtmTokenBuilder = new io.agora.rtm.RtmTokenBuilder();
        String rtmToken = "";
        try {
            rtmToken = rtmTokenBuilder.buildToken(appId, appCertificate, uid, 
                    io.agora.rtm.RtmTokenBuilder.Role.Rtm_User, (int) privilegeExpireTime);
        } catch (Exception e) {
            System.err.println("Failed to build RTM token: " + e.getMessage());
        }

        Map<String, Object> result = new HashMap<>();
        result.put("appId", appId);
        result.put("channelName", channelName);
        result.put("uid", uid);
        result.put("token", token);
        result.put("rtmToken", rtmToken);
        result.put("expiresIn", expiresIn);
        result.put("expireAt", Instant.ofEpochSecond(privilegeExpireTime).toString());

        return result;
    }
}
